crates/env/src/engine/on_chain/impls.rs

```diff
@@ -726,4 +727,40 @@ impl TypedEnvBackend for EnvInstance {
         let enc_code_hash = scope.take_encoded(code_hash);
         ext::unlock_delegate_dependency(enc_code_hash)
     }
+
+    fn xcm_execute<E, Call>(&mut self, msg: &VersionedXcm<Call>) -> Result<()>
+    where
+        E: Environment,
+        Call: scale::Encode,
+    {
+        let mut scope = self.scoped_buffer();
+
+        // Double encoding the message as the host fn expects an encoded message.
+        let enc_msg = scope.take_encoded(&scale::Encode::encode(msg));
```

Comment:
so this was because there was some double encoding going on.
You define an xcm message in ink! but it has to be passed encoded to the extrinsic 
These requirements were recently introduced to xcm but then rolled back a few days ago (https://github.com/paritytech/polkadot-sdk/pull/4266)

So I would suggest we leave it like this and when next xcm version is published we can just do
```
let enc_msg = scope.take_encoded(msg);
```